src/reconnector.ts

```diff
@@ -107,11 +125,54 @@ export class Reconnector {
     this._backoffActive = false;
   }
 
-  hardReconnect () {
-    this.client.logger.info('Attempting to reconnect with new channel.');
-    this._cleanupReconnect();
+  hardReconnect (): Promise<void> {
+    if (this._hardReconnectRetryInfo) {
+      return this._hardReconnectRetryInfo.promise;
+    }
+
+    this.cleanupReconnect();
     this._hasConnected = false;
-    return this.client.connect();
+
+    this._hardReconnectRetryInfo = {} as any;
+    this._hardReconnectRetryInfo!.promise = new Promise<void>((resolve, reject) => {
+      /* defer this to allow it to be canceled */
+      this._hardReconnectRetryInfo!.resolve = resolve;
+      this._hardReconnectRetryInfo!.reject = reject;
```

Comment:
I didn't want to defer this promise, but if we clear the interval without the promise ever completing, then calls to `hardReconnect()` could end up pending indefinitely. A pretty common scenario is if the hard reconnect interval is running, and then someone calls `client.disconnect()`. We would want to stop any pending `hardReconnect` calls. 

On the fence about whether to `resolve()` or `reject()` if someone cancels a pending `hardReconnect()`. Thoughts?

Reply:
If someone calls disconnect, we should reject with some sort of `Cancelled` error.

Reply:
👍  will do